Sources/TSCBasic/Lock.swift

```diff
@@ -163,4 +167,24 @@ public final class FileLock {
         defer { unlock() }
         return try body()
     }
+    
+    public static func withLock<T>(fileToLock: AbsolutePath, lockFilesDirectory: AbsolutePath? = nil, type: LockType = .exclusive, body: () throws -> T) throws -> T {
+        // unless specified, we use the tempDirectory to store lock files
+        let lockFilesDirectory = lockFilesDirectory ?? localFileSystem.tempDirectory
+        if !localFileSystem.exists(lockFilesDirectory) {
+            throw FileSystemError(.noEntry, lockFilesDirectory)
+        }
+        if !localFileSystem.isDirectory(lockFilesDirectory) {
+            throw FileSystemError(.notDirectory, lockFilesDirectory)
+        }
+        // use the parent path to generate unique filename in temp
+        var lockFileName = (resolveSymlinks(fileToLock.parentDirectory).appending(component: fileToLock.basename)).components.joined(separator: "_")
```

Comment:
There's somewhat of a risk of overflow here since most operating systems / file systems have a much shorter file name limit than path limit.  I don't know if that will be an issue in practice — an alternate approach to consider here might be to hash all the but the last _n_ components of the path, which should still be indicative of what the lock file refers to.

Reply:
we can also re-root the path into temp a potential solution. would that be better than what we have here now?